Adding the support for batch file annotation using a csv file

peymanvahidi opened the pull request:
The goal is to be able to use `biotope annotate batch --from-csv <csv path>` to annotate multiple files instead of going through these files one by one with the `biotope annotate interactive`.

One thing that can be done is do generate a `.biotope.csv` file when using `biotope add` with a flag, which has some default annotations already (as well as all other suggested annotations). Users can modify this csv file and then annotate their files.

Updated checklist (August, 4th)
- [x] Add `biotope status --detailed` option. (fix #7)
- [x] Insure `.git` and `.biotope` are in the same dir during initialization. (fix #9)
- [ ] Add `biotope annotate interactive -r <directory or pattern>`
- [x] `biotope annotate --to-csv` -> creates a `.biotope.csv` file, which you have your annotations created from Croissant. It would be easier for the users to update this annotation file.
- [x] Add `biotope annotate batch --from-csv`, synonymous to `biotope annotate -r --from-csv` -> uses `.biotope.csv` to update the Croissant metadata (or `biotope annotate export` or sth)
- [x] When doing `biotope add <sth>` we should have the minimum needed metadata with prefilled values.
- [x] Added `--non-interactive` flag for the `biotope init` (fix #15)
- [x] Write tests for all the new features.

peymanvahidi committed:
Adding flag for a detailed status

- Users can understand the reason of their validation fails
- Added some new tests for the new feature

peymanvahidi committed:
Adding support for annotation using relative csv file

peymanvahidi commented:
I also made the changes to have the option `biotope status --detailed`
fix #7

peymanvahidi commented on biotope/commands/annotate.py:
```diff
@@ -1240,9 +1622,23 @@ def get_staged_files(biotope_root: Path) -> list:
         for file_path in result.stdout.splitlines():
-            if file_path.startswith(".biotope/datasets/") and file_path.endswith(".jsonld"):
+            # Handle both cases: biotope project at git root and in subdirectory
+            if biotope_relative_to_git == Path("."):
+                # Biotope project is at git root
+                expected_prefix = ".biotope/datasets/"
+                metadata_file_path = file_path
+            else:
+                # Biotope project is in a subdirectory
+                expected_prefix = f"{biotope_relative_to_git}/.biotope/datasets/"
+                if file_path.startswith(str(biotope_relative_to_git) + "/"):
+                    # Strip the biotope relative path to get the path relative to biotope root
+                    metadata_file_path = file_path[len(str(biotope_relative_to_git)) + 1:]
+                else:
+                    continue
+            
+            if file_path.startswith(expected_prefix) and file_path.endswith(".jsonld"):
                 # Read the metadata file to get file information
-                metadata_file = biotope_root / file_path
+                metadata_file = biotope_root / metadata_file_path
```
Also this is the update I did to be able to handle the .biotope init in the subdirectories in a git repo.

slobentanzer replied:
I think we should prohibit `.git` and `.biotope` folders being in different directories, it may cause all manner of problems.

slobentanzer commented:
Thanks @heispv!

Proposed changes:

Use `-r` for all batch processing:
- `biotope annotate interactive -r <directory or pattern>` could be used to indicate batch annotation of multiple files (but only stating the info interactively once)
- `biotope add -r`
- `biotope annotate -r --from-csv` could be used to indicate non-interactive batch annotation (in this case from data filled in into a CSV table)
- `biotope annotate batch --from-csv` could be synonymous with `biotope annotate -r`

`biotope annotate --to-csv` command instead of `create` / `load` → create the .biotope.csv file in root to be used with the batch update
- `biotope annotate export` synonymous with --to-csv (or other formats)

Automatically pre-fill croissant json files on `biotope add` already.

peymanvahidi committed:
Insure having .biotope and .git in the same dir for initialization

- Modified related tests based on this

peymanvahidi commented:
Based on 5bf8425, `Biotope` only works if it's initialized in the same dir where `.git` lives, we also check if we're not in the git submodules.

slobentanzer commented:
@heispv thanks; why do we not support submodules, generally? if the user is knowledgeable, I think it is not a problem. as long as .biotope and .git are in the same directory, we shouldn't have conflicts, right?

The thing I wanted to prevent was that .biotope was only in a submodule of a parent .git repo without .biotope, that could be confusing. Is there any problem to be anticipated if we have this structure?
```
- .git
- .biotope
- some submodule dir
    - .git
    - .biotope
```

I think this could be the natural result of cloning one biotope project inside another biotope project, correct? Maybe we need some tests to cover this situation.

peymanvahidi committed:
Fix: git status parsing in biotope status command

- Modified files were incorrectly shown as staged due to stripped
spaces in porcelain output format parsing
- Fixed some test files as well

peymanvahidi commented:
Hi @slobentanzer  , Yes, that can be done. Maybe I misunderstood our last discussion, I thought we're going to prevent having submodules as well, making it easier for the new users. Also not sure if it makes harder for development purposes or not, but if you think this an important feature we should have, I'll remove the restriction.

slobentanzer commented:
Not necessarily an "important" feature, but also no need to prevent this case. The only thing I wanted to prevent is one git repo with multiple biotope projects in subdirectories, as that seems chaotic.

peymanvahidi commented:
Agree, I'll update it soon. 😉

peymanvahidi committed:
Allow having git submodules when initing biotope

peymanvahidi committed:
fix: use relative path for Dataset name field in add command

peymanvahidi commented:
@slobentanzer  I'm working on automatic pre-fill of croissant when doing `biotope add`.
Currently in the `validation` command it checks if we have `name, description, creator, dateCreated, distribution`, from these required fields, `name, description and distribution` will be filled automatically. To be able to pre-fill all the required fields, we should also fill the `dateCreated and creator.name`. If we are going to have a close behavior as Git, we should say the users that they should enter their name and email, like we do it in `git config --global user.name <name>` and `git config --global user.email <email>` but maybe in an interactive way, and put the dataset creation time as the current time, if they want, they can change it later.
What do you think?

Or even better, just get these info from git, and if they haven't set it using `git config --global` just tell them that they should use git's command to set them?

slobentanzer commented:
@heispv Interesting idea, we can think about that (taking the user name from git config). Generally, what is the state of this PR, should we merge or do we need to do something, still? Sorry, had to prioritise other things in the last two weeks, but happy to pick this up again.

peymanvahidi commented:
@slobentanzer No worries, I'm also a bit busy with writing my thesis, so I haven't pushed new updates for a while.
I think it's early to merge this PR yet, I'll complete all the tasks, and then we can push it.

peymanvahidi committed:
feat: prefill creator and dateCreated in biotope add metadata

- Add top-level creator field from git config user.name/email
- Set dateCreated to current timestamp for new datasets
- Show helpful tip if git identity not configured

peymanvahidi committed:
feat: add support for batch annotation editing using .biotope.csv file

peymanvahidi committed:
fix: metadata not updating based on the creator name

- Added logic to ensure creator name is captured and updated correctly from CSV, even with altered mappings.
- Improved handling of the creator object to prevent in-place mutations and preserve existing email addresses.
- Introduced tests to verify that changes in the creator column of the CSV reflect in the JSON-LD metadata.

peymanvahidi committed:
feat: add non-interactive initialization option to biotope

peymanvahidi commented:
Now when using `biotope init -n` or `biotope init --non-interactive`, the name of the project will be `<your_name_from git>_project`.
This is enough for initialization of the biotope, then users can easily add some files, and use `biotope annotate -r --to-csv` or `biotope annotate batch --to-csv` to generate the csv file from their *.jsonld, and edit the csv file, and then use `biotope annotate batch --from-csv`. This way the *.jsonld files would be updated and the intermediate `.biotope.csv` file would be removed after that.

peymanvahidi commented:
@slobentanzer Could you please check this? I believe currently we're in a good shape to merge this PR.
The only thing left is the `biotope annotate interactive -r <directory or pattern>` which we should think more about how we're going to implement this.
